web/source/utils/version.ts

```diff
@@ -0,0 +1,77 @@
+namespace com.keyman.utils {
+  // Dotted-decimal version
+  export class Version {
+    public static readonly NO_DEFAULT_KEYCAPS = new Version("12.0");
+
+    private readonly components: number[]
+
+    constructor(text: String) {
+      let parts = text.split('.');
+      let componentArray: number[] = [];
+
+      if(parts.length < 2) {
+        throw new Error("Version string must have at least a major and minor component!");
+      }
+
+      for(let i=0; i < parts.length; i++) {
+        let value = parseInt(parts[i], 10);
+        if(isNaN(value)) {
+          throw new Error("Version string components must be numerical!");
+        }
+
+        componentArray.push(value);
+      }
+
+      this.components = componentArray;
+    }
+
+    get major(): number {
+      return this.components[0];
+    }
+
+    get minor(): number {
+      return this.components[1];
+    }
+
+    toString(): string {
+      return this.components.join('.');
+    }
+
+    equals(other: Version): boolean {
+      if(this.components.length != other.components.length) {
+        return false;
+      } else {
+        for(let i=0; i < this.components.length; i++) {
+          if(this.components[i] != other.components[i]) {
+            return false;
+          }
+        }
+
+        return true;
+      }
+    }
+
+    precedes(other: Version): boolean {
+      // If the version info depth differs, we need a flag to indicate whether or not 'this' instance wins ties.
+      // 12.0 is considered to precede 12.0.0.
```

Comment:
I think `12.0.0` = `12.0`. `12.0.1` would succeed `12.0.0`. But missing should be the same as zero.

Reply:
👍

I wasn't sure exactly how we wanted this handled, but figured this would easily be enough to spark the discussion.  Can do.

Reply:
Comment reflects old behaviour. Code looks correct.